crates/rrg/src/action/yara_scan_memory.rs

```diff
@@ -0,0 +1,564 @@
+    Scan(yara_x::ScanError),
+}
+
+impl std::fmt::Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "failed to scan process memory: ")?;
+        match self {
+            Error::OpenProcessMemory(error) => {
+                write!(f, "failed to open process memory for reading: {}", error)
+            }
+            Error::Scan(error) => {
+                write!(f, "error during scanning: {}", error)
+            }
+        }
+    }
+}
+
+impl std::error::Error for Error {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match self {
+            Error::OpenProcessMemory(error) => Some(error),
+            Error::Scan(scan_error) => Some(scan_error),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct ErrorItem {
+    /// PID of the process that we encountered an error scanning.
+    pid: u32,
+    /// Actual error that occurred when scanning `pid`.
+    error: Error,
+}
+
+#[derive(Debug)]
+struct OkItem {
+    /// PID of the process this item belongs to.
+    pid: u32,
+    /// YARA rules that matched when scanning this process' memory.
+    matching_rules: Vec<Rule>,
+}
+
+/// Result of the `yara_scan_memory` action for one single process.
+type Item = Result<OkItem, ErrorItem>;
+
+impl crate::response::Item for Item {
+    type Proto = rrg_proto::yara_scan_memory::Result;
+
+    fn into_proto(self) -> Self::Proto {
+        let mut proto = Self::Proto::new();
+        match self {
+            Err(ErrorItem { pid, error }) => {
+                proto.set_pid(pid);
+                proto.set_error(error.to_string());
+            }
+            Ok(OkItem {
+                pid,
+                matching_rules,
+            }) => {
+                proto.set_pid(pid);
+                proto.set_matching_rules(matching_rules.into_iter().map(Into::into).collect());
+            }
+        }
+        proto
+    }
+}
+
+/// Reads and scans a single region of memory, breaking it up into chunks of `chunk_size` with an overlap of
+/// `chunk_overlap`. Only returns an error if scanning failed; read errors are ignored.
+fn scan_region<M: MemoryReader>(
+    region: &MappedRegion,
+    scanner: &mut Scanner,
+    memory: &mut M,
+    chunk_size: u64,
+    chunk_overlap: u64,
+) -> Result<(), Error> {
+    let mut offset = region.start_address();
+    while offset < region.end_address() {
+        let remaining = region.end_address() - offset;
+        let length = remaining.min(chunk_size + chunk_overlap);
+        // Any process will most likely have at least one region
+        // that cannot be read succesfully, so there's no point
+        // in reporting an error to the user if that happens,
+        // just ignore it and continue.
+        if let Ok(buf) = memory.read_chunk(offset, length) {
+            scanner.scan(offset as usize, &buf).map_err(Error::Scan)?;
+        }
+        offset = offset.saturating_add(chunk_size);
```

Comment:
I presume we use `.saturating_add` to avoid (a rather theoretical) overflow, but then aren't we setting ourselves up for (a rather theoretical) infinite loop?

Reply:
> since the loop condition is `offset < region.end_address()` even in the case where `region.end_address() == u64::MAX` the loop will terminate when the add saturates.

Ah, yep, you are right.